model/task_queue_service_dependency.go

```diff
@@ -0,0 +1,375 @@
+	if item, ok := t.nodeItemMap[id]; ok {
+		return item
+	}
+
+	return nil
+}
+
+func (t *basicCachedDAGDispatcherImpl) getNodeByItemID(id string) graph.Node {
+	if node, ok := t.itemNodeMap[id]; ok {
+		return node
+	}
+
+	return nil
+}
+
+func (t *basicCachedDAGDispatcherImpl) addEdge(from string, to string) error {
+	fromNodeID := t.itemNodeMap[from].ID()
+	toNodeID := t.itemNodeMap[to].ID()
+
+	// Cannot add a self edge!
+	if fromNodeID == toNodeID {
+		grip.Alert(message.Fields{
+			"dispatcher": "dependency-task-dispatcher",
+			"function":   "addEdge",
+			"message":    "cannot add a self edge to a Node",
+			"node_id":    fromNodeID,
+		})
+	}
+
+	edge := simple.Edge{
+		F: simple.Node(fromNodeID),
+		T: simple.Node(toNodeID),
+	}
+	t.graph.SetEdge(edge)
+
+	return nil
+}
+
+func (t *basicCachedDAGDispatcherImpl) rebuild(items []TaskQueueItem) error {
+	for i := range items {
+		// Add each individual <TaskQueueItem> node to the graph
+		t.addItem(&items[i])
+	}
+
+	// Save the task groups
+	t.taskGroups = map[string]schedulableUnit{}
+	for _, item := range items {
+		if item.Group != "" {
+			// If it's the first time encountering the task group create an entry for it in the taskGroups map.
+			// Otherwise, append to the taskQueueItem array in the map.
+			id := compositeGroupId(item.Group, item.BuildVariant, item.Project, item.Version)
+			if _, ok := t.taskGroups[id]; !ok {
+				t.taskGroups[id] = schedulableUnit{
+					id:       id,
+					group:    item.Group,
+					project:  item.Project,
+					version:  item.Version,
+					variant:  item.BuildVariant,
+					maxHosts: item.GroupMaxHosts,
+					tasks:    []TaskQueueItem{item},
+				}
+			} else {
+				taskGroup := t.taskGroups[id]
+				taskGroup.tasks = append(taskGroup.tasks, item)
+				t.taskGroups[id] = taskGroup
+			}
+		}
+	}
+
+	// Add edges for task dependencies
+	for _, item := range items {
+		for _, dependency := range item.Dependencies {
+			if err := t.addEdge(item.Id, dependency); err != nil {
+				return errors.Wrapf(err, "Failed to create in-memory task queue of TaskQueueItems for distro '%s'; error defining a DirectedGraph incorporating task dependencies", t.distroID)
+			}
+		}
+	}
+
+	// Sort the graph. Use a lexical sort to resolve ambiguities, because node order is the
+	// order that we received these in.
+	sorted, err := topo.SortStabilized(t.graph, lexicalReversed)
+	if err != nil {
+		grip.Alert(message.WrapError(err, message.Fields{
+			"dispatcher": "dependency-task-dispatcher",
+			"function":   "rebuild",
+			"message":    "problem sorting tasks within the DirectedGraph",
+		}))
+	}
+	t.sorted = sorted
+	t.lastUpdated = time.Now()
+
+	return nil
+}
+
+type byIDReversed []graph.Node
+
+func (n byIDReversed) Len() int           { return len(n) }
+func (n byIDReversed) Less(i, j int) bool { return n[i].ID() > n[j].ID() }
+func (n byIDReversed) Swap(i, j int)      { n[i], n[j] = n[j], n[i] }
+func lexicalReversed(nodes []graph.Node) {
```

Comment:
I don't think that this function helps a lot, and I might remove it, but it's fine